### ISSUE TITLE
Is there a way start app in another culture rather than the default?

### ISSUE THREAD

**Issue by fabercs** [haestflod/asp.net-core-mvc-localized-routing#29]
Hi,

First of all, thanks for your efforts on this.

I am trying to implement your work on my project. I have two supported languages, English and Turkish. The default culture should be Turkish (tr), however I would like to keep controller and action names in English in the source code. These names would generate the urls in English as well. So in Startup.cs I changed default culture to > `tr`
`LocalizationRouteDataHandler.DefaultCulture = "tr";`

However, it does not generate localized urls, instead generates in english. But if I set ti to `en`, and then change the culture to `tr`, then it generate the localized urls.

What I mean is if I give an example on your sample project, I want this project to start with
`LocalizationRouteDataHandler.DefaultCulture = "fi";`
and all actions shold be localized on startup.

Could you help me on this?

**Comment by haestflod** [comment]
I did not think about the use case where you would want a controller in english but a different default culture when I wrote this. 

In the `LocalizationRouteConvention.cs` where it sets the actual routing templates it uses the ControllerName for the default culture. 
In the function `AddControllerRoutes()` it does the following:
```
string controllerName = controllerModel.ControllerName;

LocalizationRouteDataHandler.AddControllerRouteData(controllerName, LocalizationRouteDataHandler.DefaultCulture, controllerName);
```
And the `ControllerName` is the class name minus Controller. If `HomeController` then the ControllerName is `Home`

To support what you'd want I think instead of using the `ControllerModel.ControllerName` directly it would need to use a localized name, I'm not sure how difficult it would be to implement without breaking existing functionality. Or if more changes are necessary with how it changes the `<a href="">`

**Comment by fabercs** [comment]
The only option I came up without changing any core functionality (as I am not familiar with the routing in net core) is at the begginnig set the DefaultCulture to `en` then as soon as request reach to HomeController, reset DefaultCulture to `tr` or `fi` which is desired as default. :) I did not tried that but I will. 

But, no matter what, this codebase should support this usecase as well, any supported culture could be default (think of accept headers).

**Comment by haestflod** [comment]
I agree that it should be supported. It should be a common use case since coding in english is common.

Most likely the routing because of default ASP.NET will keep the controller name and actions, I don't remember if you can remove that.

So I think most likely for you for HomeController the routing would end up being:
/Home,
/Ev   (Google translate Home to turkish)
/en/Home

And then it's the job of the `CultureActionLinkTagHelper.cs` to make sure that it selects
/Ev and not /Home if both exists.

**Comment by fabercs** [comment]
Actually what I need is;


> /en/Home/Index
> /tr/Ev/Giris

> /en/Example/Parameter
> /tr/Ornek/Parametre

I tried below, but the default culture's routes are being omitted in somewhere, I guess... Also I tried to use `Route` attributes `Name` property for a workaround to rename actions in Turkish.

```
// Set up cultures
LocalizationRouteDataHandler.DefaultCulture = "tr";
LocalizationRouteDataHandler.SupportedCultures = new Dictionary<string, string>()
{
    { "en", "English" },
    { "tr", "Türkçe" }
};
```

```
[LocalizationRoute("tr", "Ornek")]
[LocalizationRoute("en", "Example")]
public class ExampleController : LocalizationController
{
    [LocalizationRoute("tr", "Parametre")]
    [LocalizationRoute("en", "Parameter")]
    public IActionResult Parameter(int index, string test)
    {
        ViewData["index"] = index;
        ViewData["test"] = test;
        ViewData["post"] = false;
        return View();
    }
}
```
It take long time to grasp the code if this is doable, what do you think?

**Comment by haestflod** [comment]
If I remember correctly myself they are omitted. I can't remember there being an option to force the default culture. 

If you want to force /en/ and /tr/ you could try and create a third culture and redirect to /tr/ if that third culture is found by using a middleware for requests.

I tested some and if you define a `[LocalizationRoute]` for a default culture you can still access it by the /defaultCulture/name route but every route that doesn't have a `[LocalizationRoute]` will throw an error for default culture.

Which means that it's probably easier to fix the initial issue, That you can write the controller name in english but use a different default culture.

**Comment by fabercs** [comment]
> If you want to force /en/ and /tr/ you could try and create a third culture and redirect to /tr/ if that third culture is found by using a middleware for requests as a workaround.

I want to do this but, shouldn't `UseMvc` be the last middleware in the pipeline? Lets say I have a dummy default culture `wa`, at the end of the pipleine I need to redierct reuqest to `/tr/Home`, how can I do this?

**Comment by haestflod** [comment]
You just use a middleware I believe. 
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-2.2

It could also be possible inside the `LocalizationController.cs`- `OnActionExecuting`
To add a check for culture and maybe return a `context.Result = new RedirectResult()`

**Comment by dkadd** [comment]
> So I think most likely for you for HomeController the routing would end up being:
> /Home,
> /Ev (Google translate Home to turkish)
> /en/Home

Can you explain what changes do i need to make for this.

"
